### ISSUE TITLE
Python support for channels from file descriptors

### ISSUE THREAD

**Issue by fahhem** [grpc/grpc#23760]
### Is your feature request related to a problem? Please describe.

We need a way for a client to connect to a server that is potentially behind a firewall. Our solution is to have the server initiate the TCP connection, but would then like to pass that establish connection into a grpc Channel to then use grpc as expected. We use Python for the client and server, otherwise this would be possible in C++ using existing functionality ([CreateInsecureChannelFromFd](https://github.com/grpc/grpc/blob/b73f7acb344c606123a77e7ce7bcb4c7ebe6fcb4/src/cpp/client/create_channel_posix.cc))

### Describe the solution you'd like

Add support into the Python [insecure_channel](https://github.com/grpc/grpc/blob/089d90933a778b9fad65c36848b2843c253cb152/src/python/grpcio/grpc/__init__.py#L1905) function to take a file descriptor instead of a target string (or another `insecure_channel_from_fd` function). That would then create a Cython-based [Channel](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi#L441) from the file descriptor instead of `bytes target` and that would call [`grpc_insecure_channel_create_from_fd`](https://github.com/grpc/grpc/blob/f2f8bbd0080e9d4d61e50e1d361bb8bc761a32b4/src/core/ext/transport/chttp2/client/insecure/channel_create_posix.cc#L38)

That change to those 2 files would be sufficient. We're happy to make the changes if the grpc community would be welcoming of the changes.

### Describe alternatives you've considered

We could rip out the Channel classes into our codebase and modify them there (we have to ship this to a customer so we can't just branch grpc). That's not great, we would rather share this work with others.

### Additional context

There's a go version of this for the exact same reason here: https://github.com/dustin-decker/grpc-firewall-bypass

Thank you!

**Comment by maged** [comment]
It would also be nice to get this into `secure_channel` if possible.

**Comment by gnossen** [comment]
This sounds totally reasonable. Thank you for offering to do the legwork here. One thing that will have to happen to get this into the public API is to do a quick write-up (gRFC) for the changes since they're an addition to the public API. ([details on that](https://github.com/grpc/proposal)).

I'd be happy to be your point of contact and help you with any question / troubles with CI you might have.

**Comment by fahhem** [comment]
No, we moved on from this. We did find the solution was possible in c++ so it was only a matter of exposing more functionality

**Comment by Mosseridan** [comment]
> 

@fahhem Is you're solution open-source?
Do you have any good resources you can point me to?
Thanks.

**Comment by fahhem** [comment]
We didn't implement anything, so nothing to open source, sorry. No particular resources either

**Comment by tarlano** [comment]
Hi Fahrzin

Please can we reopen the issue and then you can expose the functionality using your c++?

Thank you so much
Anthony
________________________________

**Comment by fahhem** [comment]
I don't have the ability to close or re-open issues, nor do I have any C++ or Python code that's not already in grpc. If you need this, please open a new issue and ping actual grpc developers, not me

**Comment by nimrod-gileadi** [comment]
I'm also interested in this feature. I want to use gRPC for communicating with a Python interpreter, sandboxed using [Sandbox2](https://developers.google.com/code-sandboxing/sandbox2). Sandbox2 offers file descriptors as a way of communicating with the host process, and it would be good to be able to start gRPC clients or servers over those file descriptor. It seems possible to do this in C++ (using `CreateInsecureChannelFromFd`), but not in Python.

**Comment by XuanWang-Amos** [comment]
This is still a valid use case, currently we don't have enough bandwidth to implement it, we can help guide the whole process if anyone is interested to implement this feature. And like mentioned above, we'll also need a gRFC for this.
